### ISSUE TITLE
Support 1.7.x

### ISSUE THREAD

**Issue by coolacid** [ammaraskar/pyCraft#33]
Would be supper helpful to support 1.7.x for existing Modded servers.

**Comment by dkkline** [comment]
There are some issues with modded servers.

I believe some mods still introduce new packets, even though they should all use the plugin message system. This would mean we'd have to introduce new packet definitions for each and every mod not using the plugin message system. I think @ammaraskar would agree, that it is beyond the scope of pyCraft to support mods not using the standardized packet definitions (plugin message 0x17).

**Comment by ammaraskar** [comment]
>This would mean we'd have to introduce new packet definitions for each and every mod not using the plugin message system

Not necessarily, if there's a packet we don't support we just skip it.
However, I really don't feel like back porting the code for whatever packets 1.7 used, I'd rather keep 1.8 as the base for the rewrite.

**Comment by dkkline** [comment]
We would still be unable to parse the non-notchian packets though.

I agree that backporting to 1.7 seems a bit silly. Mods will eventually update to 1.8, and from then on we can support each version simply by adding git tags.

**Comment by ammaraskar** [comment]
>We would still be unable to parse the non-notchian packets though.

They still have to send the packet header meaning we know exactly how many bytes to skip reading, which is already handled here: https://github.com/ammaraskar/pyCraft/blob/master/minecraft/networking/connection.py#L234-L241

**Comment by Mati0703** [comment]
I made 1.7 support for this changing some code in packets.py and looking on wiki.vg. I will publish this when finished.

**Comment by ProgrammerDan** [comment]
Thanks Mati, your efforts were exactly what I needed. Trying to verify hotpatch for the 0-day using the exploit for a 1.7.10 server; took your efforts and adapted to protocol 5, with the exploit as an optional command after start: https://github.com/ProgrammerDan/pyCraft/tree/1.7.10-exploit

**Comment by Mati0703** [comment]
I tried to make expilot working with 1.7 but i missed something on blockplace packet, so it disconnect from server. I dont have much time to look at it :C

**Comment by ProgrammerDan** [comment]
Having the same problem, getting frame width exceptions. Something with the
NBT packet data, I think. Digging deeper atm.

**Comment by ProgrammerDan** [comment]
@Mati0703 Check the prior linked branch. It's working now. Trick was dealing with NBT in 1.7.10 -- it must be GZIP (in this context, have to use gzip lib, not zlib), and you have to prepend the NBT with a SHORT indicating size of gzip'd data. I've confirmed on a spigot 1.7.10 unpatched server; it brings it down. Might need to send a few exploit packets, but the failure is progressive -- at first seems to just be a lag spike, but after a few minutes, server blows up with stack trace failures and dies. Haven't tried vanilla 1.7.10, but I'd image the same result. Should be easy to adapt from here to earlier versions as you'll just have to tweak individual packets in the reactors, and the exploit should still work as written.

**Comment by Mati0703** [comment]
@ProgrammerDan Thanks! You solved this! Now i can check if my 1.7.2 patched server can handle this :)

**Comment by Mati0703** [comment]
Simply patched in 1.7.2 server by adding to NTBTagList this:

```
            if ( (j+1) > 2097152) {
            	System.out.println("PyCraft Hack detected!");
                throw new RuntimeException("Tried to read NBT tag that was too big; tried to allocate: " + (j+1) + "bytes where max allowed: 2097152");
            }
```

in void load(DataInput datainput, int i) :)

**Comment by ammaraskar** [comment]
Thanks for the work on porting pyCraft over to 1.7.x guys!
